prov/gni/src/gnix_fabric.c

```diff
@@ -484,44 +449,61 @@ static int gnix_getinfo(uint32_t version, const char *node, const char *service,
 		}
 	}
 
+	ret = __gnix_getinfo_resolve_node(node, service, flags, gnix_info);
+	if (ret != FI_SUCCESS)
+		goto  err;
+
 	gnix_info->mode = mode;
 	gnix_info->fabric_attr->name = strdup(gnix_fab_name);
 	gnix_info->tx_attr->caps = gnix_info->caps;
 	gnix_info->tx_attr->mode = gnix_info->mode;
 	gnix_info->rx_attr->caps = gnix_info->caps;
 	gnix_info->rx_attr->mode = gnix_info->mode;
 
-	if (ep_type_unspec) {
-		struct fi_info *dg_info = fi_dupinfo(gnix_info);
+	*info = gnix_info;
 
-		if (!dg_info) {
-			GNIX_WARN(FI_LOG_FABRIC, "cannot copy info\n");
-			goto err;
-		}
+	GNIX_DEBUG(FI_LOG_FABRIC, "Returning EP type: %s\n",
+		   fi_tostr(&ep_type, FI_TYPE_EP_TYPE));
+	return FI_SUCCESS;
+err:
+	_gnix_freeinfo(gnix_info);
+	return -FI_ENODATA;
+}
 
-		dg_info->ep_attr->type = FI_EP_DGRAM;
-		gnix_info->next = dg_info;
+static int gnix_getinfo(uint32_t version, const char *node, const char *service,
+			uint64_t flags, struct fi_info *hints,
+			struct fi_info **info)
+{
+	int ret = 0;
+	struct fi_info *info_ptr;
+
+	/* Note that info entries are added to the head of 'info', that is,
+	 * they are preferred in the reverse order shown here. */
+
+	*info = NULL;
+
+	ret = _gnix_ep_getinfo(FI_EP_MSG, version, node, service, flags,
+			       hints, &info_ptr);
+	if (ret == FI_SUCCESS) {
+		info_ptr->next = *info;
+		*info = info_ptr;
 	}
 
-	*info = gnix_info;
+	ret = _gnix_ep_getinfo(FI_EP_DGRAM, version, node, service, flags,
+			       hints, &info_ptr);
+	if (ret == FI_SUCCESS) {
+		info_ptr->next = *info;
+		*info = info_ptr;
+	}
 
-	return 0;
-err:
-	if (gnix_info) {
-		if (gnix_info->tx_attr) free(gnix_info->tx_attr);
-		if (gnix_info->rx_attr) free(gnix_info->rx_attr);
-		if (gnix_info->ep_attr) free(gnix_info->ep_attr);
-		if (gnix_info->domain_attr) free(gnix_info->domain_attr);
-		if (gnix_info->fabric_attr) free(gnix_info->fabric_attr);
-		free(gnix_info);
+	ret = _gnix_ep_getinfo(FI_EP_RDM, version, node, service, flags,
+			       hints, &info_ptr);
+	if (ret == FI_SUCCESS) {
+		info_ptr->next = *info;
+		*info = info_ptr;
```

Comment:
If ep_attr->type is specified, only the matching EP type fi_info would be returned.